react/components/EXPERIMENTAL_Table/hooks/useTableBulkActions.tsx

```diff
@@ -0,0 +1,230 @@
+    const headerRender = () => {
+      const selectedRowsLength = bulkState.selectedRows.length
+      const itemsLength = bulkedItems.length
+      const isChecked = selectedRowsLength === itemsLength
+      const isPartial =
+        selectedRowsLength > 0 && selectedRowsLength < itemsLength
+
+      return (
+        <Checkbox
+          checked={isChecked}
+          onClick={selectAllVisibleRows}
+          id="all"
+          partial={isPartial}
+        />
+      )
+    }
+
+    const cellRender = ({ rowData }) => (
+      <Checkbox
+        checked={bulkState.selectedRows.some(row => row.id === rowData.id)}
+        onClick={() => selectRow(rowData)}
+        id={`${rowData.id}`}
+        disabled={bulkState.allLinesSelected}
+      />
+    )
+
+    return hasBulkActions
+      ? [
+          {
+            id: 'bulk',
+            width: 40,
+            headerRender,
+            cellRender,
+          },
+          ...columns,
+        ]
+      : columns
+  }, [bulkState.selectedRows, bulkState.allLinesSelected])
+
+  useEffect(() => {
+    if (bulkActions && bulkActions.onChange) {
+      const selectedParameters = bulkState.allLinesSelected
+        ? { allLinesSelected: true }
+        : { selectedRows: bulkState.selectedRows }
+      bulkActions.onChange(selectedParameters)
+    }
+  }, [bulkState.selectedRows, bulkState.allLinesSelected, bulkActions])
+
+  const selectAllRows = useCallback(
+    () =>
+      dispatch({
+        type: 'SELECT_ALL_ROWS',
+        selectedRows: bulkedItems,
+      }),
+    [bulkState.selectedRows]
+  )
+
+  const deselectAllRows = useCallback(
+    () => dispatch({ type: 'DESELECT_ALL_ROWS' }),
+    [bulkState.selectedRows]
+  )
+
+  const selectRow = useCallback(
+    (row: BulkedItem) => dispatch({ type: 'SELECT_ROW', row }),
+    [bulkState.selectedRows]
+  )
+
+  const setSelectedRows = useCallback(
+    (selectedRows: Array<BulkedItem>) =>
+      dispatch({ type: 'SET_SELECTED_ROWS', selectedRows }),
+    [bulkState.selectedRows]
+  )
+
+  const setAllLinesSelected = useCallback(
+    (allLinesSelected: boolean) =>
+      dispatch({ type: 'SET_ALL_LINES_SELECTED', allLinesSelected }),
+    [bulkState.allLinesSelected]
+  )
+
+  const selectAllVisibleRows = useCallback(
+    () =>
+      bulkState.selectedRows.length <= bulkedItems.length &&
+      bulkState.selectedRows.length !== 0
+        ? deselectAllRows()
+        : setSelectedRows(bulkedItems),
+    [bulkState.selectedRows, bulkState.allLinesSelected]
+  )
+
+  return {
+    hasBulkActions,
+    hasPrimaryBulkAction,
+    hasSecondaryBulkActions,
+
+    /** state */
+    bulkState,
+
+    /** data */
+    bulkedColumns,
+    bulkedItems,
+
+    /** handler fn */
+    selectAllRows,
+    deselectAllRows,
+    selectAllVisibleRows,
+    selectRow,
+    setSelectedRows,
+    setAllLinesSelected,
+  }
+}
+
+function reducer(state: BulkState, action: Action) {
+  switch (action.type) {
+    case 'SET_SELECTED_ROWS': {
```

Comment:
Those strings could be `const`s since they're used several times.

Reply:
Solved with `enum` type